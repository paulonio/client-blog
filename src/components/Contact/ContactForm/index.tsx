'use client';

import React from 'react';
import emailjs from '@emailjs/browser';
import { useForm, SubmitHandler } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';
import { useTranslation } from 'next-i18next';

import Button from '@/components/Button';
import SelectArrowIcon from '@/icons/SelectArrowIcon';

import { CATEGORIES } from '@/constants/constants';
import { NAME_REGEXP, EMAIL_REGEXP } from '@/constants/contact';

import { inter } from '@/styles/fonts';

import styles from './styles.module.scss';
import { stringToKey } from '@/utils/utils';

interface FormType {
  fullName: string;
  email: string;
  topic: string;
  message: string;
}

const schema = yup.object({
  fullName: yup
    .string()
    .trim()
    .min(5)
    .matches(NAME_REGEXP, 'name should contain only latin letters and at least 2 words')
    .required(),
  email: yup.string().trim().email().matches(EMAIL_REGEXP, 'email must be a valid').required(),
  topic: yup.string().oneOf(CATEGORIES).required(),
  message: yup.string().trim().min(5).max(100).required(),
});

const ContactForm = () => {
  const { t } = useTranslation(['contact', 'common']);
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<FormType>({ resolver: yupResolver(schema) });

  const onSubmit: SubmitHandler<FormType> = (data) => {
    emailjs
      .send('service_zijuo46', 'template_b5ggbe9', { ...data }, 'WYxkO50AUT5PWYUGB')
      .then(() => reset());
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} className={styles.form}>
      <input
        className={`${styles.input} ${inter.className}`}
        type="text"
        placeholder={t('form.name')}
        {...register('fullName')}
        data-testid="name"
      />
      <p className={styles.form__error}>{errors.fullName?.message}</p>
      <input
        className={`${styles.input} ${inter.className}`}
        type="text"
        placeholder="Email"
        {...register('email')}
        data-testid="email"
      />
      <p className={styles.form__error}>{errors.email?.message}</p>
      <div className={styles.select__wrapper}>
        <select
          className={`${styles.select} ${inter.className}`}
          defaultValue="topic"
          {...register('topic')}
        >
          <option disabled value="topic">
            {t('form.topic')}
          </option>
          {CATEGORIES.map((category) => (
            <option key={category} className={inter.className} value={category}>
              {t(stringToKey(category), { ns: 'common' })}
            </option>
          ))}
        </select>
        <SelectArrowIcon className={styles.arrow} />
      </div>
      <p className={styles.form__error}>{errors.topic?.message}</p>
      <textarea
        className={`${styles.textarea} ${inter.className}`}
        placeholder={t('form.message')}
        {...register('message')}
        data-testid="message"
      />
      <p className={styles.form__error}>{errors.message?.message}</p>
      <Button type="submit">{t('form.send')}</Button>
    </form>
  );
};

export default ContactForm;
