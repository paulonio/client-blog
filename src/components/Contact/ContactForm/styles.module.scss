@import '../../../styles/sass/constants';

.form {
  margin-top: 32px;
  width: 100%;
  max-width: 768px;
  display: flex;
  flex-direction: column;

  @media screen and (max-width: 768px) {
    padding: 0 10px;
  }

  &__error {
    height: 21px;
    padding-left: 8px;
    font-size: 16px;
    line-height: 21px;
    color: red;
  }
}

.input, 
.textarea, 
.select {
  padding: 24px;
  font-size: 16px;
  line-height: 28px;
  font-weight: 400;
  color: $color-black;
  border: solid 1px rgba($color-dark-s, 0.5);
  outline: none;

  &::placeholder {
    color: $color-black;
    opacity: 0.9;
  }
}

.textarea {
  height: 144px;
  resize: none;

  &::placeholder {
    font-size: 16px;
    line-height: 28px;
    font-weight: 400;
    color: $color-black;
    opacity: 1;
  }
}

.select {
  width: 100%;
  background-color: $color-white;
  appearance: none;

  &__wrapper {
    position: relative;
  }

  &:hover {
    cursor: pointer;
  }
}

.arrow {
  position: absolute;
  top: 36px;
  right: 63px;
  transform: translate(-50%);
  pointer-events: none;

  @media screen and (max-width: 768px) {
    right: 32px;
  }
}