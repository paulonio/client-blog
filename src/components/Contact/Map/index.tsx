'use client';

import React, { FC, useEffect, useRef } from 'react';

interface MapProps {
  center: google.maps.LatLngLiteral;
  zoom: number;
}

const Map: FC<MapProps> = ({ center, zoom }) => {
  const mapRef = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    if (mapRef.current) {
      new window.google.maps.Map(mapRef.current, {
        center,
        zoom,
      });
    }
  });

  return <div ref={mapRef} id="map" />;
};

export default Map;
