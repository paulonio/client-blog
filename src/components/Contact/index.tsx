'use client';

import React, { useMemo } from 'react';
import { Wrapper } from '@googlemaps/react-wrapper';

import ContactForm from './ContactForm';
import Map from './Map';

import { sen } from '@/styles/fonts';

import styles from './styled.module.scss';

const Contact = () => {
  const center = useMemo(() => ({ lat: 18.52043, lng: 73.856743 }), []);

  return (
    <section className={styles.contact__wrapper}>
      <div className={styles.heading__wrapper}>
        <h5>CONTACT US</h5>
        <h1 className={sen.className}>Letâ€™s Start a Conversation</h1>
        <p className="body-s">
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt
          ut labore et dolore magna aliqua. Ut enim ad minim.
        </p>
      </div>
      <div className={styles.info__wrapper}>
        <div>
          <p className={`${styles.info__heading}  body-l`}>Working hours</p>
          <h4 className={sen.className}>Monday To Friday</h4>
          <h4 className={sen.className}>9:00 AM to 8:00 PM</h4>
          <p className="body-s">Our Support Team is available 24/7</p>
        </div>
        <div>
          <p className={`${styles.info__heading} body-l`}>Contact Us</p>
          <h4 className={sen.className}>020 7993 2905</h4>
          <p className="body-s">hello@finsweet.com</p>
        </div>
      </div>
      <ContactForm />
      <Wrapper apiKey="YOUR_API_KEY">
        <Map center={center} zoom={4} />
      </Wrapper>
    </section>
  );
};

export default Contact;
