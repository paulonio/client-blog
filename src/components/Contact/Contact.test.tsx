import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import Contact from '.';
import ContactForm from './ContactForm';

const TEST_VALUE = 'test';

describe('Contact', () => {
  it('should render contact page', () => {
    const { container } = render(<Contact />);

    expect(container).toMatchSnapshot();
  });

  it('should render contact form', () => {
    const { container } = render(<ContactForm />);

    expect(container).toMatchSnapshot();
  });

  it('should correctly change inputs', () => {
    render(<ContactForm />);

    const name: HTMLInputElement = screen.getByTestId('name');
    const email: HTMLInputElement = screen.getByTestId('email');
    const message: HTMLTextAreaElement = screen.getByTestId('message');
    fireEvent.change(name, { target: { value: TEST_VALUE } });
    fireEvent.change(email, { target: { value: TEST_VALUE } });
    fireEvent.change(message, { target: { value: TEST_VALUE } });
    expect(name.value).toEqual(TEST_VALUE);
    expect(email.value).toEqual(TEST_VALUE);
    expect(message.value).toEqual(TEST_VALUE);
  });
});
