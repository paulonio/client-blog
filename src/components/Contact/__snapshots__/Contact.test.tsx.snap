// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact should render contact form 1`] = `
<div>
  <form
    class="form"
  >
    <input
      class="input className"
      data-testid="name"
      name="fullName"
      placeholder="form.name"
      type="text"
    />
    <p
      class="form__error"
    />
    <input
      class="input className"
      data-testid="email"
      name="email"
      placeholder="Email"
      type="text"
    />
    <p
      class="form__error"
    />
    <div
      class="select__wrapper"
    >
      <select
        class="select className"
        name="topic"
      >
        <option
          disabled=""
          selected=""
          value="topic"
        >
          form.topic
        </option>
        <option
          class="className"
          value="Business"
        >
          business
        </option>
        <option
          class="className"
          value="Economy"
        >
          economy
        </option>
        <option
          class="className"
          value="Startup"
        >
          startup
        </option>
        <option
          class="className"
          value="Technology"
        >
          technology
        </option>
      </select>
      <svg
        class="arrow"
        fill="currentColor"
        focusable="false"
        height="8"
        role="img"
        viewBox="0 0 12 8"
        width="12"
      >
        <path
          d="M2 2L6 6L10 2"
          stroke="black"
          stroke-linecap="square"
          stroke-width="2"
        />
      </svg>
    </div>
    <p
      class="form__error"
    />
    <textarea
      class="textarea className"
      data-testid="message"
      name="message"
      placeholder="form.message"
    />
    <p
      class="form__error"
    />
    <button
      class="button"
      type="submit"
    >
      form.send
    </button>
  </form>
</div>
`;

exports[`Contact should render contact page 1`] = `
<div>
  <section>
    <div
      class="contact__wrapper"
    >
      <div
        class="heading__wrapper"
      >
        <h5>
          heading.subtitle
        </h5>
        <h1
          class="className"
        >
          heading.title
        </h1>
        <p
          class="body-s"
        >
          heading.body
        </p>
      </div>
      <div
        class="info__wrapper"
      >
        <div>
          <p
            class="info__heading  body-l"
          >
            info.working
          </p>
          <h4
            class="className"
          >
            info.days
          </h4>
          <h4
            class="className"
          >
            info.hours
          </h4>
          <p
            class="body-s"
          >
            info.support
          </p>
        </div>
        <div>
          <p
            class="info__heading body-l"
          >
            info.contact
          </p>
          <h4
            class="className"
          >
            020 7993 2905
          </h4>
          <p
            class="body-s"
          >
            hello@finsweet.com
          </p>
        </div>
      </div>
      <form
        class="form"
      >
        <input
          class="input className"
          data-testid="name"
          name="fullName"
          placeholder="form.name"
          type="text"
        />
        <p
          class="form__error"
        />
        <input
          class="input className"
          data-testid="email"
          name="email"
          placeholder="Email"
          type="text"
        />
        <p
          class="form__error"
        />
        <div
          class="select__wrapper"
        >
          <select
            class="select className"
            name="topic"
          >
            <option
              disabled=""
              selected=""
              value="topic"
            >
              form.topic
            </option>
            <option
              class="className"
              value="Business"
            >
              business
            </option>
            <option
              class="className"
              value="Economy"
            >
              economy
            </option>
            <option
              class="className"
              value="Startup"
            >
              startup
            </option>
            <option
              class="className"
              value="Technology"
            >
              technology
            </option>
          </select>
          <svg
            class="arrow"
            fill="currentColor"
            focusable="false"
            height="8"
            role="img"
            viewBox="0 0 12 8"
            width="12"
          >
            <path
              d="M2 2L6 6L10 2"
              stroke="black"
              stroke-linecap="square"
              stroke-width="2"
            />
          </svg>
        </div>
        <p
          class="form__error"
        />
        <textarea
          class="textarea className"
          data-testid="message"
          name="message"
          placeholder="form.message"
        />
        <p
          class="form__error"
        />
        <button
          class="button"
          type="submit"
        >
          form.send
        </button>
      </form>
    </div>
    <div
      class="map__wrapper"
    >
      <iframe
        allowfullscreen=""
        height="100%"
        loading="lazy"
        referrerpolicy="no-referrer-when-downgrade"
        src="https://www.google.com/maps/d/u/0/embed?mid=1UndmFwsUO1yidEWv7GjURvLv5E9UOrM&ehbc=2E312F"
        style="border: 0px;"
        title="map"
        width="100%"
      />
    </div>
  </section>
</div>
`;
